sync2/multipeer/multipeer_test.go

```diff
@@ -0,0 +1,372 @@
+			// delegate to the real fullsync
+			return mt.reconciler.FullSync(ctx, peers)
+		})
+	mt.syncBase.EXPECT().Derive(gomock.Any()).DoAndReturn(func(p p2p.Peer) multipeer.PeerSyncer {
+		mt.mtx.Lock()
+		defer mt.mtx.Unlock()
+		require.Contains(mt, pl.get(), p)
+		s := NewMockPeerSyncer(mt.ctrl)
+		s.EXPECT().Peer().Return(p).AnyTimes()
+		// TODO: do better job at tracking Release() calls
+		s.EXPECT().Release().AnyTimes()
+		expSync := s.EXPECT().Sync(gomock.Any(), gomock.Nil(), gomock.Nil())
+		if numFails != 0 {
+			expSync.Return(errors.New("sync failed"))
+			numFails--
+		}
+		return s
+	}).Times(times)
+}
+
+// satisfy waits until all the expected mocked calls are made.
+func (mt *multiPeerSyncTester) satisfy() {
+	require.Eventually(mt, func() bool {
+		mt.mtx.Lock()
+		defer mt.mtx.Unlock()
+		return mt.ctrl.Satisfied()
+	}, time.Second, time.Millisecond)
+}
+
+func TestMultiPeerSync(t *testing.T) {
+	const numSyncs = 3
+
+	t.Run("split sync", func(t *testing.T) {
+		mt := newMultiPeerSyncTester(t, 0)
+		ctx := mt.start()
+		mt.clock.BlockUntilContext(ctx, 1)
+		// Advance by sync interval. No peers yet
+		mt.clock.Advance(time.Minute)
+		mt.clock.BlockUntilContext(ctx, 1)
+		// It is safe to do EXPECT() calls while the MultiPeerReconciler is blocked
+		mt.addPeers(10)
+		// Advance by peer wait time. After that, 6 peers will be selected
+		// randomly and probed
+		mt.syncBase.EXPECT().Count().Return(50, nil).AnyTimes()
+		for i := 0; i < numSyncs; i++ {
+			plSplit := mt.expectProbe(6, rangesync.ProbeResult{
+				FP:    "foo",
+				Count: 100,
+				Sim:   0.5, // too low for full sync
+			})
+			mt.syncRunner.EXPECT().SplitSync(gomock.Any(), gomock.Any()).DoAndReturn(
+				func(_ context.Context, peers []p2p.Peer) error {
+					require.ElementsMatch(t, plSplit.get(), peers)
+					return nil
+				})
+			mt.syncBase.EXPECT().Wait()
+			mt.clock.BlockUntilContext(ctx, 1)
+			plFull := mt.expectProbe(6, rangesync.ProbeResult{
+				FP:    "foo",
+				Count: 100,
+				Sim:   1, // after sync
+			})
+			mt.expectFullSync(plFull, 6, 0)
+			mt.syncBase.EXPECT().Wait()
+			if i > 0 {
+				mt.clock.Advance(time.Minute)
+			} else if i < numSyncs-1 {
+				mt.clock.Advance(10 * time.Second)
+			}
+			mt.satisfy()
+		}
+		mt.syncBase.EXPECT().Wait()
+	})
+
+	t.Run("full sync", func(t *testing.T) {
+		mt := newMultiPeerSyncTester(t, 10)
+		mt.syncBase.EXPECT().Count().Return(100, nil).AnyTimes()
+		require.False(t, mt.reconciler.Synced())
+		var ctx context.Context
+		for i := 0; i < numSyncs; i++ {
+			pl := mt.expectProbe(6, rangesync.ProbeResult{
+				FP:    "foo",
+				Count: 100,
+				Sim:   0.99, // high enough for full sync
+			})
+			mt.expectFullSync(pl, 6, 0)
+			mt.syncBase.EXPECT().Wait()
+			if i == 0 {
+				//nolint:fatcontext
+				ctx = mt.start()
+			} else {
+				// first full sync happens immediately
+				mt.clock.Advance(time.Minute)
+			}
+			mt.clock.BlockUntilContext(ctx, 1)
+			mt.satisfy()
+		}
```

Comment:
If the first loop behaves differently would it make sense to indicate it as such more clearly? This should also get rid of the linter warning.

```suggestion
		pl := mt.expectProbe(6, rangesync.ProbeResult{
			FP:    "foo",
			Count: 100,
			Sim:   0.99, // high enough for full sync
		})
		mt.expectFullSync(pl, 6, 0)
		mt.syncBase.EXPECT().Wait()
		mt.clock.Advance(time.Minute) // first sync happens immediatly
		mt.clock.BlockUntilContext(context.Background(), 1)
		mt.satisfy()

		for i := 1; i < numSyncs; i++ {
			pl := mt.expectProbe(6, rangesync.ProbeResult{
				FP:    "foo",
				Count: 100,
				Sim:   0.99, // high enough for full sync
			})
			mt.expectFullSync(pl, 6, 0)
			mt.syncBase.EXPECT().Wait()
			ctx := mt.start()
			mt.clock.BlockUntilContext(ctx, 1)
			mt.satisfy()
		}
```

Reply:
It's the other way around, we need to do `mt.start()` in the initial iteration and advance the clocks in the following ones. But otherwise that's probably the right idea, so I updated the code, except that I wrapped expectations in the nested func `expect` to highlight the fact that the expectations are the same right after startup and when it's time to do more syncs